File: AR7ExportToolbar/ToolStripExport/Exporters/SectionReportExporter.cs
using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Linq;
using System.Text;
using GrapeCity.ActiveReports.Document;
using GrapeCity.ActiveReports.Export;

namespace AR7ExportToolbar.ToolStripExport.Exporters
{
	public abstract class SectionReportExporter : ISectionReportExporter
	{
		private static string GetAssemblyVersion()
		{
			return typeof(SectionDocument).Assembly.GetName().Version.ToString();
		}

		private readonly string FullTypeName;

		protected SectionReportExporter(string type, string assembly) : this(type, assembly, GetAssemblyVersion())
		{
		}

		protected SectionReportExporter(string type, string assembly, string version)
		{
			const string FullTypeNameFormat = "{0}, {1}, Version={2}, Culture=neutral, PublicKeyToken=cc4967777c49a3ff";

			FullTypeName = string.Format(FullTypeNameFormat, type, assembly, version);
		}

		protected Type CreateType()
		{
			return Type.GetType(FullTypeName, false, true);
		}

		public bool IsAvailable()
		{
			var exportType = CreateType();

			return (exportType != null);
		}

		public abstract void Export(SectionDocument document, string filename, NameValueCollection settings);

		public virtual IDocumentExportEx GetExporter()
		{
			var type = CreateType();
			var exp = Activator.CreateInstance(type) as IDocumentExportEx;

			return exp;
		}

		public abstract string DefaultExtension
		{
			get;
		}

		public abstract string FileDialogFilter
		{
			get;
		}
	}
}

File: AR7ExportToolbarVB/ToolStripExport/Exporters/SectionReportExporter.vb
Imports GrapeCity.ActiveReports.Document
Imports System.Collections.Specialized
Imports GrapeCity.ActiveReports.Export

Public MustInherit Class SectionReportExporter
	Implements ISectionReportExporter

	Private Shared Function GetAssemblyVersion() As String
		Return GetType(SectionDocument).Assembly.GetName().Version.ToString()
	End Function

	Private ReadOnly FullTypeName As String

	Protected Sub New(ByVal type As String, ByVal assembly As String)
		Me.New(type, assembly, GetAssemblyVersion())
	End Sub

	Protected Sub New(ByVal type As String, ByVal assembly As String, ByVal version As String)
		Const FullTypeNameFormat As String = "{0}, {1}, Version={2}, Culture=neutral, PublicKeyToken=cc4967777c49a3ff"

		FullTypeName = String.Format(FullTypeNameFormat, type, assembly, version)
	End Sub

	Protected Function CreateType() As Type
		Return Type.GetType(FullTypeName, False, False)
	End Function

	Public Function IsAvailable() As Boolean _
		Implements ISectionReportExporter.IsAvailable
		Dim exportType = CreateType()

		Return Not (exportType Is Nothing)
	End Function

	Public MustOverride Sub Export(ByVal document As SectionDocument, filename As String, settings As NameValueCollection)

	Public Overridable Function GetExporter() As IDocumentExportEx _
		Implements ISectionReportExporter.GetExporter

		Dim type = CreateType()
		Dim exp = CType(Activator.CreateInstance(type), IDocumentExportEx)

		Return exp
	End Function

	Public MustOverride ReadOnly Property DefaultExtension As String Implements ISectionReportExporter.DefaultExtension
	Public MustOverride ReadOnly Property FileDialogFilter As String Implements ISectionReportExporter.FileDialogFilter

End Class
